doc/alps_submitter.md

```diff
@@ -0,0 +1,237 @@
+Feature Expr -> Semantic Analyze  -> Feature Columns Code Generation -> Estimator
+
+### Feature Expressions 
+In SQLFlow, we use Feature Expressions to represent the feature engineering process and convert it into the code snippet using [TensorFlow Feature Column API](https://www.tensorflow.org/guide/feature_columns).
+
+<b>Feature Expressions</b>
+```
+NUMERIC(key, dtype, shape)
+
+BUCKETIZED(source_column, boundaries)
+
+CATEGORICAL_IDENTITY(key, num_buckets, default_value)
+
+CATEGORICAL_HASH(key, hash_bucket_size, dtype)
+
+CATEGORICAL_VOCABULARY_LIST(key, vocabulary_list, dtype, default_value, num_oov_buckets)
+
+CATEGORICAL_VOCABULARY_FILE(key, vocabulary_file, vocabulary_size, num_oov_buckets, default_value, dtype)
+
+CROSS(keys, hash_bucket_size, hash_key)
+```
+
+> The feature expressions must exist in `COLUMN` block.
+
+Here is an example which do `BUCKETIZED` on `c2` then `CROSS` with `c1`.
+
+```sql
+select 
+    c1, c2, c3 as class
+from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+  ...
+COLUMN
+  CROSS([NUMERIC(c1), BUCKETIZED(NUMERIC(c2), [0, 10, 100])])
+LABEL class
+
+```
+
+### Semantic Analyze
+Feature Expressions except for Tensorflow Feature Column API should raise an error.
+```sql
+/* Not supported */
+select * from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+    ...
+COLUMN
+    NUMERIC(f1 * 10)
+```
+
+### Feature Columns Code Generation
+We transform feature columns expression to a code snippet and wrap it as a `CustomFCBuilder` which extends from `alps.feature.FeatureColumnsBuilder`.
+
+Review the above example, the generated code snippet is this:
+
+```python
+from alps.feature import FeatureColumnsBuilder
+
+class CustomFCBuilder(FeatureColumnsBuilder):
+
+  def build_feature_columns():
+    fc1 = tf.feature_column.numeric_column('c1')
+    fc2 = tf.feature_column.numeric_column('c2')
+    fc3 = tf.feature_column.bucketized_column(fc2, boundaries = [0, 10, 100])
+    fc4 = tf.feature_column.crossed_column([fc2, fc3])
+    
+    return [fc4]
+```
+
+ALPS framework will execute this code snippet and pass the result to the constructor method of the estimator.
+
+### Parameters
+
+We use `WITH` block to set the parameters of training.
+
+If the name is prefixed with `estimator`, it is the parameter of the constructor method of the `Estimator`.
+
+If the name is prefixed with `train_spec`, it is the parameter of the constructor method of the `TrainSpec`.
+
+If the name is prefixed with `input_fn`, it is the parameter of the `input_fn`.
+
+Let's create a DNNClassifier example, the minimum parameters of the constructor method are `hidden_units` and `feature_columns`.
+
+```sql
+select 
+    c1, c2, c3 as class
+from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+    estimator.hidden_units = [10, 20],
+    train_spec.max_steps = 2000,
+    input_fn.batch_size = 512
+COLUMN
+    CROSS([NUMERIC(c1), BUCKETIZED(NUMERIC(c2), [0, 10, 100])])
+LABEL class
+...
+```
+
```

Comment:
How about changing the following sentence to

## Future Development

For now, we will combine the `feature_columns` as one dense tensor and feed it to the model. If we want to support models with multiple inputs, we will need an `AS` keyword annotate different inputs.